include/odp/api/spec/comp.h

```diff
@@ -0,0 +1,603 @@
+ *
+ * @param	session		Session handle
+ *
+ * @retval	0 on success
+ * @retval	<0 on failure
+ */
+int odp_comp_session_destroy(odp_comp_session_t session);
+
+/**
+ * Compression set dictionary
+ *
+ * Sets the dictionary value to the corresponding compression session.
+ *
+ * @param	session		Session handle
+ * @param	dict		Pointer to dictionary.
+
+ * @retval 0 on success
+ * @retval <0 on failure
+ *
+ * @note: Application should check dictionary support using
+ * odp_comp_alg_capability()
+ */
+int odp_comp_set_dict(odp_comp_session_t session, odp_comp_dict_t *dict);
+
+/**
+ * Synchronous packet compression operation
+ *
+ * This operation does packet compression in synchronous mode. A successful operation
+ * returns the number of successfully processed input packets and updates the
+ * results in the corresponding output packets. Outputted packets contain
+ * compression results metadata (odp_comp_packet_result_t), which should be
+ * checked for operation status.
+ *
+ * For every input packet entry in 'pkt_in' array, application should pass
+ * corresponding valid output packet handle. If any error occurs during
+ * processing of packets, the API returns with number of entries successfully
+ * processed. When hashing is configured along with compression operation the
+ * result is appended at the end of the output packet.
+ *
+ * Processed data length can be computed by subtracting 'digest_len' from
+ * output_data_range.len where 'digest_len' can be queried from
+ * odp_comp_hash_alg_capability().
+ *
+ * @param          pkt_in	Packets to be processed
+ * @param          pkt_out	Packet handle array for resulting packets
+ * @param          num_pkt	Number of packets to be processed
+ * @param          param	Operation parameters
+ *
+ * @return Number of input packets consumed (1 ... num_pkt)
+ * @retval <0 on failure
+ *
+ * @note Both 'pkt_in' and 'pkt_out' array should be of same length, Results
+ * are undefined if otherwise.
+ */
+int odp_comp_op(const odp_packet_t pkt_in[], odp_packet_t pkt_out[],
+		int num_pkt, const odp_comp_packet_op_param_t param[]);
+
+/**
+ * Asynchronous packet compression  operation
+ *
+ * This operation does packet compression in asynchronous mode. It processes
+ * packets otherwise identical to odp_comp_op(), but the resulting packets are
+ * enqueued to 'compl_queue' configured during session (odp_comp_session_t)
+ * creation. For every input packet entry in in_pkt array, user should pass
+ * corresponding valid output packet handle. On return, API returns with
+ * number of entries successfully submitted for operation.
+ *
+ * In case of partially accepted array i.e.
+ * when number of packets returned < num_pkt, application may attempt to
+ * resubmit subsequent entries via calling any of the operation API.
+ *
+ * Processed data length can be computed by subtracting 'digest_len' from
+ * output_data_range.len where 'digest_len' can be queried from
+ * odp_comp_hash_alg_capability().
```

Comment:
Isn't the digest only the output from compress operations? Decompress operations should simply produce the uncompressed data expanded from the compressed input. What is the application expected to do with the hash? Presumably the compressed data + digest should be passed to whoever is doing to do the decompress so that it can serve as an integrity check? Does this mean we should have a defined error return that indicates the decompress operation was successful but the hash didn't compare?

Reply:
Digest output is available during both compress and decompresses operation.
During decompression, the Hashing validation is done by the application since the compression engine does not have the hash on uncompressed data to validate.